java-frontend/src/main/java/org/sonar/java/se/xproc/MethodYield.java

```diff
@@ -93,54 +93,49 @@ public MethodYield(ExplodedGraph.Node node, MethodBehavior behavior) {
-      return parametersConstraints[index];
+      return parametersConstraints.get(index);
     }
     return null;
   }
 
-  private static Set<ProgramState> programStatesForConstraint(Collection<ProgramState> states, SymbolicValue invokedArg, Constraint constraint) {
-    Set<ProgramState> programStates = new LinkedHashSet<>();
-    if (constraint instanceof ObjectConstraint) {
-      ObjectConstraint objectConstraint = (ObjectConstraint) constraint;
-      states.forEach(state -> programStates.addAll(invokedArg.setConstraint(state, objectConstraint)));
-    } else if (constraint instanceof BooleanConstraint) {
-      BooleanConstraint booleanConstraint = (BooleanConstraint) constraint;
-      states.forEach(state -> programStates.addAll(invokedArg.setConstraint(state, booleanConstraint)));
-    }
-    return programStates;
-  }
-
   /**
    * For varArgs methods, only apply the constraint on single array parameter, in order to not 
    * wrongly apply it on all the elements of the array.
    */
   private boolean applicableOnVarArgs(int index, List<Type> types) {
-    if (index < parametersConstraints.length - 1) {
+    if (index < parametersConstraints.size() - 1) {
       // not the varArg argument
       return true;
     }
-    if (parametersConstraints.length != types.size()) {
+    if (parametersConstraints.size() != types.size()) {
       // more than one element in the variadic part
       return false;
     }
     Type argumentType = types.get(index);
     return argumentType.isArray() || argumentType.is("<nulltype>");
   }
 
-  public Constraint[] parametersConstraints() {
-    return parametersConstraints;
-  }
+  private static Set<ProgramState> programStatesForConstraint(Collection<ProgramState> states, SymbolicValue invokedArg,
+                                                              PMap<Class<? extends Constraint>, Constraint> constraints) {
+    Set<ProgramState> programStates = new LinkedHashSet<>(states);
 
-  @CheckForNull
-  public Constraint parameterConstraint(int parameterIndex) {
-    return parametersConstraints[parameterIndex];
+    constraints.forEach((d, c) ->  {
+      Set<ProgramState> newPs = new LinkedHashSet<>();
+      for (ProgramState programState : programStates) {
+        newPs.addAll(invokedArg.setConstraint(programState, c));
+      }
+      programStates.clear();
+      programStates.addAll(newPs);
+    });
+    return programStates;
   }
 
   public void setParameterConstraint(int index, @Nullable Constraint constraint) {
-    Preconditions.checkArgument(index < parametersConstraints.length);
-    parametersConstraints[index] = constraint;
+    Preconditions.checkArgument(index < parametersConstraints.size());
+    PMap<Class<? extends Constraint>, Constraint> constraintsForParam = parametersConstraints.get(index);
+    parametersConstraints.add(index, constraintsForParam.put(constraint.getClass(), constraint));
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') NullPointerException might be thrown as 'constraint' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)